src/app/OperationalDeviceManager.cpp

```diff
@@ -0,0 +1,181 @@
+
+struct ConnectionErrorContext
+{
+    chip::OperationalDeviceProxy * device;
+    CHIP_ERROR error;
+};
+
+} // namespace
+
+namespace chip {
+
+CHIP_ERROR OperationalDeviceManager::AcquireDevice(PeerId peerId, Callback::Callback<OnDeviceConnected> * onConnection,
+                                                   Callback::Callback<OnDeviceConnectionFailure> * onFailure)
+{
+    Dnssd::ResolvedNodeData resolutionData;
+
+    bool nodeIDWasResolved = (mConfig.dnsCache != nullptr && mConfig.dnsCache->Lookup(peerId, resolutionData) == CHIP_NO_ERROR);
+
+    OperationalDeviceProxy * device = FindActiveDevice(peerId);
+    if (device == nullptr)
+    {
+        if (nodeIDWasResolved)
+        {
+            device = mConfig.devicePool->Allocate(mConfig.sessionInitParams, peerId, resolutionData);
+        }
+        else
+        {
+            device = mConfig.devicePool->Allocate(mConfig.sessionInitParams, peerId);
+        }
+
+        if (device == nullptr)
+        {
+            onFailure->mCall(onFailure->mContext, nullptr, peerId, CHIP_ERROR_NO_MEMORY);
+            return CHIP_ERROR_NO_MEMORY;
+        }
+        device->EnqueueConnectionCallbacks(&mOnConnectedCallback, &mOnConnectionFailureCallback);
+        return device->Connect(onConnection, onFailure, mConfig.dnsResolver);
+    }
+    else
+    {
+        OperationalDeviceProxy * newDevice = mConfig.devicePool->Allocate(mConfig.sessionInitParams, peerId);
+        if (newDevice == nullptr)
+        {
+            onFailure->mCall(onFailure->mContext, nullptr, peerId, CHIP_ERROR_NO_MEMORY);
+            return CHIP_ERROR_NO_MEMORY;
+        }
+        // Try to reuse the session
+        if (device->IsConnected())
+        {
+            newDevice->SetConnectedSession(device->GetSecureSession().Value());
+            onConnection->mCall(onConnection->mContext, newDevice);
+        }
+        // Otherwise wait for the outgoing CASESession to be connected
+    }
+    return CHIP_NO_ERROR;
+}
+
+void OperationalDeviceManager::ReleaseDevice(OperationalDeviceProxy * device)
+{
+    if (device != nullptr)
+    {
+        mConfig.devicePool->Release(device);
+    }
+}
+
+static Loop NotifyDeviceConnected(const void * context, OperationalDeviceProxy * device)
+{
+    const OperationalDeviceProxy * connectedDevice = static_cast<const OperationalDeviceProxy *>(context);
+    if (device != connectedDevice && !device->IsActive() && device->GetPeerId() == connectedDevice->GetPeerId())
+    {
+        device->SetConnectedSession(connectedDevice->GetSecureSession().Value());
+    }
+    return Loop::Continue;
+}
+
+void OperationalDeviceManager::HandleDeviceConnected(void * context, OperationalDeviceProxy * device)
+{
+    OperationalDeviceManager * manager = static_cast<OperationalDeviceManager *>(context);
+    if (!manager->mNotifyingConnectionSucess)
+    {
+        manager->mNotifyingConnectionSucess = true;
+        manager->mConfig.devicePool->ForEachDevice(device, NotifyDeviceConnected);
```

Comment:
Why is this notifying every single active object? Shouldn't it just be notifying device's that 'match'?